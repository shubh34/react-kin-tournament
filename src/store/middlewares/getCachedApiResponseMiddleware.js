/* eslint-disable import/no-mutable-exports */
import { isRSAA, RSAA } from 'redux-api-middleware';
import head from 'lodash/head';

import { API_CACHE_TIMER } from '../../configs/config';


export const getCachedApiResponseMiddleware = () => () => next => (action) => {
	if (!isRSAA(action)) {
		return next(action);
	}
	const cachedApiResponseString = sessionStorage.getItem(head(action[RSAA].types));
	if (cachedApiResponseString) {
		const cachedApiResponse = JSON.parse(cachedApiResponseString);
		if (Date.now() - cachedApiResponse.createdTimeStamp < API_CACHE_TIMER) {
			const nextAction = {
				[RSAA]: {
					...action[RSAA],
					fetch: async () => new Response(JSON.stringify(cachedApiResponse.response),
						{
							status: 200,
							headers: {
								'Content-Type': 'application/json',
							},
						}),
				},
			};
			return next(nextAction);
		}
		sessionStorage.clear(action[RSAA].endpoint);
	}

	return next(action);
};
