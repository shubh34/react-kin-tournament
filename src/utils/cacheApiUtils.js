export const toBeCachedApiResponse = ['GET_TOURNAMENT_LEADERS_REQUEST', 'GET_TOURNAMENT_PRIZES_REQUEST'];

export const cacheApiResponse = (response) => {
	const { meta: { cacheKey = '' }, payload } = response;
	if (toBeCachedApiResponse.includes(cacheKey) && !sessionStorage.getItem(cacheKey)) {
		const responseToBeCached = {
			createdTimeStamp: Date.now(),
			response: payload,
		};
		sessionStorage.setItem(cacheKey, JSON.stringify(responseToBeCached));
	}
};
