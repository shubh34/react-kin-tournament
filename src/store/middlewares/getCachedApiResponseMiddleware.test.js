import { RSAA } from 'redux-api-middleware';
import { getCachedApiResponseMiddleware } from './getCachedApiResponseMiddleware';
import { cacheApiResponse } from '../../utils/cacheApiUtils';


const setup = (
	next = jest.fn(),
) => ({
	next,
	middleware: getCachedApiResponseMiddleware()()(next),
});

describe('getCachedApiResponseMiddleware', () => {
	it('calls next with original action if not RSAA action', () => {
		const { middleware, next } = setup();
		const action = {
			type: 'DEMO_TEST',
		};

		middleware(action);
		expect(next).toBeCalledWith(action);
	});

	it('should call the api when the response is not stored in session storage', () => {
		const { middleware, next } = setup();
		const response = {
			meta: { cacheKey: 'GET_TOURNAMENT_LEADERS_REQUEST' },
			payload: { test: 'new Response' },
		};
		cacheApiResponse(response);
		const action = {
			[RSAA]: {
				endpoint: 'test/RSAA',
				types: ['GET_TOURNAMENT_LEADERS_REQUEST'],
			},

		};

		middleware(action);

		expect(next).toMatchSnapshot();
	});
	it('should call the the api when the response is in session storage', () => {
		const { middleware, next } = setup();

		const action = {
			[RSAA]: {
				endpoint: 'test/RSAA',
			},
		};

		middleware(action);

		expect(next).toBeCalledWith(
			expect.objectContaining({
				[RSAA]: {
					endpoint: 'test/RSAA',
				},
			}),
		);
	});
	it('should call the api when api cached createdTimeStamp is more than API_CACHE_TIMER', () => {
		const { middleware, next } = setup();

		const action = {
			[RSAA]: {
				endpoint: 'test/RSAA',
				types: ['GET_TOURNAMENT_PRIZES_REQUEST'],
			},
		};
		const apiResponse = {
			createdTimeStamp: '156673552',
			response: {
				test: 'test',
			},
		};
		sessionStorage.setItem('GET_TOURNAMENT_PRIZES_REQUEST', JSON.stringify(apiResponse));

		middleware(action);

		expect(next).toBeCalledWith(
			expect.objectContaining({
				[RSAA]: {
					endpoint: 'test/RSAA',
					types: ['GET_TOURNAMENT_PRIZES_REQUEST'],
				},

			}),
		);
    });
    
    it('should clear the sessionStorage api cached createdTimeStamp is more than API_CACHE_TIMER', () => {
		const { middleware, next } = setup();

		const action = {
			[RSAA]: {
				endpoint: 'test/RSAA',
				types: ['GET_TOURNAMENT_PRIZES_REQUEST'],
			},
		};
		const apiResponse = {
			createdTimeStamp: '156673552',
			response: {
				test: 'test',
			},
		};
		sessionStorage.setItem('GET_TOURNAMENT_PRIZES_REQUEST', JSON.stringify(apiResponse));

		middleware(action);

		expect(sessionStorage).toMatchSnapshot();
	});
});
