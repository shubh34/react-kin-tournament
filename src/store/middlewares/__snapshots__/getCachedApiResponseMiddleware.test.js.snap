// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCachedApiResponseMiddleware should call the api when the response is not stored in session storage 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "@@redux-api-middleware/RSAA": Object {
          "endpoint": "test/RSAA",
          "fetch": [Function],
          "types": Array [
            "GET_TOURNAMENT_LEADERS_REQUEST",
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`getCachedApiResponseMiddleware should clear the sessionStorage api cached createdTimeStamp is more than API_CACHE_TIMER 1`] = `Storage {}`;
